core/unix/os.c

```diff
@@ -7834,13 +7834,28 @@ mmap_check_for_module_overlap(app_pc base, size_t size, bool readable, uint64 in
                  * Note, if it is a header of a different module, then we'll not have
                  * an overlap, so we will not hit this case.
                  */
-                ASSERT_CURIOSITY(
-                    ma->start + ma->os_data.alignment ==
-                    base
-                        /* On Mac we walk the dyld module list before the
-                         * address space, so we often hit modules we already
-                         * know about. */
-                        IF_MACOS(|| !dynamo_initialized && ma->start == base));
+#ifdef AARCH64
+                /* For AArch64 systems with 64K pages it is possible for small modules
+                 * (less than 1 page of disk space) to have multiple RW LOADable data
+                 * segments. Since these segments are mapped from a single disk space they
+                 * will all have a elf_header satisfying the check above.
+                 */
+                if (ma->os_data.alignment == 65536) {
+                    size_t offset = base - ma->start;
+                    size_t seg_id = offset / ma->os_data.alignment;
```

Comment:
Some ELF files have different alignments for different segments.  I don't think we can assume they are all identical.

Reply:
From my understanding, for every module only one alignment is possible (i.e., `module_area_t.os_module_data_t.alignment`). So, the alignment between all LOAD segments should be the same. Other types of segments can surely have other alignments but these segments are not included in `module_area_t.os_module_data.segments`.

Reply:
> From my understanding, for every module only one alignment is possible

I don't think this is the case: the ELF spec does not say the every segment must have the same alignment.

Reply:
Yep not for every segment but for every segment in the segment array in `module_area_t.os_module_data.segments ` only one alignment seems to be possible. Isn't that the case? If it's not the case, should't alignment be part of the `module_segment_t`. My point is based on the DR codebase rather than pointing to an ELF specification. Based on the spec, every segment can have its own alignment as specified in the program header. But it seems that LOAD segments seem to always have the same alignment (in most cases at least). Having LOAD segments with different alignments seem to be causing issues in some settings (e.g., https://github.com/numpy/numpy/issues/16677).

Reply:
Maybe we can add a comment in the code of this PR that mentions the assumption that was made based on the current DR code (in case `os_module_data_t` changes) or change the code and move the alignment inside `module_segment_t` so that every segment can have its own alignment (and make no assumptions here about a unified alignment for the segments in the array). As a third option, we can traverse here all segments without alignment assumptions and ignore the assumptions made elsewhere in the DR code.

Reply:
The computation "seg_id = offset / ma->os_data.alignment" has more problems than just assuming constant alignment: segments are not always adjacent.  There are sometimes gaps in between.  Taking a later segment's base and subtracting the first segment's base does not tell you how many segments are in between.  I think the check here should just do a traversal and abandon the idea of trying to compute the index from the address.  It'll be inside a DOCHECK or whatnot so it won't affect release build performance.

Reply:
Great point. I completely missed the scenario of non-adjacent segments. There is also a field `contiguous` in the `os_module_data_t` struct which is meant for this scenario. So, I will do the traversal instead.

Reply:
Re: DR module code assuming single segment alignment: it does seem like that is the case.  Not ideal since it doesn't seem hard to handle different alignments, but I guess we've lived with it all this time.  There are other areas where DR doesn't do the right thing wrt segment alignment: #4737.

Reply:
Should we create a new issue explicitly reporting this single segment alignment assumption, or add some comment in the code to point it out?

Reply:
However you want to do it -- I see a curiosity assert in module_add_segment_data so we would find out violations in debug build at least.

Reply:
I added a comment in the code. Good thing that there is also a check in module_add_segment_data.